src/Datastore/Operation.php

```diff
@@ -0,0 +1,659 @@
+     *           If provided, an instance of that class will be returned instead of Entity.
+     *           If not set, {@see Google\Cloud\Datastore\Entity} will be used.
+     *     @type array $excludeFromIndexes A list of entity keys to exclude from
+     *           datastore indexes.
+     * }
+     * @return Entity
+     */
+    public function entity($key, array $entity = [], array $options = [])
+    {
+        $options += [
+            'className' => null
+        ];
+
+        if (!is_string($key) && !($key instanceof Key)) {
+            throw new InvalidArgumentException(
+                '$key must be an instance of Key or a string'
+            );
+        }
+
+        if (is_string($key)) {
+            $key = $this->key($key);
+        }
+
+        $className = $options['className'];
+        if (!is_null($className) && !is_subclass_of($className, Entity::class)) {
+            throw new InvalidArgumentException(sprintf(
+                'Given classname %s is not a subclass of Entity',
+                $className
+            ));
+        }
+
+        if (is_null($className)) {
+            $className = Entity::class;
+        }
+
+        return new $className($key, $entity, $options);
+    }
+
+    /**
+     * Allocate available IDs to a set of keys
+     *
+     * Keys MUST be in an incomplete state (i.e. including a kind but not an ID
+     * or name in their final pathElement).
+     *
+     * This method will execute a service request.
+     *
+     * @see https://cloud.google.com/datastore/reference/rest/v1/projects/allocateIds allocateIds
+     *
+     * @param Key[] $keys The incomplete keys.
+     * @param array $options Configuration options.
+     * @return Key[]
+     */
+    public function allocateIds(array $keys, array $options = [])
+    {
+        // Validate the given keys. First check types, then state of each.
+        // The API will throw a 400 if the key is complete, but it's an easy
+        // check we can handle before going to the API to save a request.
+        // @todo replace with json schema
+        $this->validateBatch($keys, Key::class, function ($key) {
+            if ($key->state() !== Key::STATE_INCOMPLETE) {
+                throw new InvalidArgumentException(sprintf(
+                    'Given $key is in an invalid state. Can only allocate IDs for incomplete keys. ' .
+                    'Given path was %s',
+                    (string) $key
+                ));
+            }
+        });
+
+        $res = $this->connection->allocateIds([
+            'projectId' => $this->projectId,
+            'keys' => $keys
+        ] + $options);
+
+        if (isset($res['keys'])) {
+            foreach ($res['keys'] as $index => $key) {
+                $end = end($key['path']);
+                $id = $end['id'];
+                $keys[$index]->setLastElementIdentifier($id);
+            }
+        }
+
+        return $keys;
+    }
+
+    /**
+     * Lookup records by key
```

Comment:
It may be documented elsewhere, so apologies if I missed it - but could we more clearly document how the user can work with the returned result in the case of missing/deferred?